source/server/listener_manager_impl.cc

```diff
@@ -410,56 +427,129 @@ void ListenerImpl::addFilterChainForApplicationProtocols(
     }
   }
 }
 
 void ListenerImpl::addFilterChainForSourceTypes(
     SourceTypesArray& source_types_array,
     const envoy::api::v2::listener::FilterChainMatch_ConnectionSourceType source_type,
+    const std::vector<std::string>& source_ips,
+    const Protobuf::RepeatedField<Protobuf::uint32>& source_ports,
+    const Network::FilterChainSharedPtr& filter_chain) {
+  if (source_ips.empty()) {
+    addFilterChainForSourceIPs(source_types_array[source_type].first, EMPTY_STRING, source_ports,
+                               filter_chain);
+  } else {
+    for (const auto& source_ip : source_ips) {
+      addFilterChainForSourceIPs(source_types_array[source_type].first, source_ip, source_ports,
+                                 filter_chain);
+    }
+  }
+}
+
+void ListenerImpl::addFilterChainForSourceIPs(
+    SourceIPsMap& source_ips_map, const std::string& source_ip,
+    const Protobuf::RepeatedField<Protobuf::uint32>& source_ports,
     const Network::FilterChainSharedPtr& filter_chain) {
-  if (source_types_array[source_type] != nullptr) {
+  if (source_ports.empty()) {
+    addFilterChainForSourcePorts(source_ips_map[source_ip], 0, filter_chain);
+  } else {
+    for (auto source_port : source_ports) {
+      addFilterChainForSourcePorts(source_ips_map[source_ip], source_port, filter_chain);
+    }
+  }
+}
+
+void ListenerImpl::addFilterChainForSourcePorts(SourcePortsMapSharedPtr& source_ports_map_ptr,
+                                                uint32_t source_port,
+                                                const Network::FilterChainSharedPtr& filter_chain) {
+  if (source_ports_map_ptr == nullptr) {
+    source_ports_map_ptr = std::make_shared<SourcePortsMap>();
+  }
+  auto& source_ports_map = *source_ports_map_ptr;
+
+  if (!source_ports_map.try_emplace(source_port, filter_chain).second) {
     // If we got here and found already configured branch, then it means that this FilterChainMatch
     // is a duplicate, and that there is some overlap in the repeated fields with already processed
     // FilterChainMatches.
     throw EnvoyException(fmt::format("error adding listener '{}': multiple filter chains with "
                                      "overlapping matching rules are defined",
                                      address_->asString()));
   }
-  source_types_array[source_type] = filter_chain;
 }
 
-void ListenerImpl::convertDestinationIPsMapToTrie() {
+namespace {
+
+// Template function for creating a CIDR list entry for either source or destination address.
+template <class T>
+std::pair<T, std::vector<Network::Address::CidrRange>> makeCidrListEntry(const std::string& cidr,
+                                                                         const T& data) {
+  std::vector<Network::Address::CidrRange> subnets;
+  if (cidr == EMPTY_STRING) {
+    if (Network::Address::ipFamilySupported(AF_INET)) {
+      static const std::string catch_all_ipv4 = "0.0.0.0/0";
```

Comment:
Static initialization order fiasco?

Reply:
I think this is only initialized the first time this code is run so I don't think so?

Reply:
Yeah, that works around the initialization order fiasco, but there is a closely related destruction order problem (https://isocpp.org/wiki/faq/ctors#construct-on-first-use-v2). Maybe add CIDR equivalents of `Utility::getIpv4AnyAddress()` etc.

Reply:
Strict adherence to this rule continues to feel like busy work, but will fix.